### ISSUE TITLE
Understand impacts of new readLine() interface on k-nucleotide benchmarks

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#20785]
When we switched to the new readLine() interface in #19937, we took a performance hit to k-nucleotide versions that switched to the new interface as seen in https://chapel-lang.org/perf/chapcs/?startdate=2022/05/02&enddate=2022/10/06&graphs=knucleotideshootoutbenchmark,submittedknucleotideshootoutbenchmark  The submitted versions initially took a hit, but then this went away when we reverted the change to them in order to have them continue to reflect the live versions on the CLBG website.

These new versions also took a code size hit (arguably a productivity hit?) in that cases where we needed to use one readLine() call per loop before now require two.

For both these reasons, I've been hesitant to have us remove the deprecated forms, at least without an understanding of why the slips happened and an agreement that we're OK with them / prefer the new interfaces despite these impacts.

**Comment by mppf** [comment]
I wonder if the performance hit has to do with the use of `mark()` in the new versions or if it has to do with calling the function to read 1 byte vs. a C function to get a whole line.

**Comment by bradcray** [comment]
Yeah, checking the code again, I think Elliot's right:  That before we accepted an array + offset version as two separate arguments, so could just start reading from that offset; and now we only take a complete array, so have to pass in a slice.  But I think he's also right that we may not have verified that to be the cause—it just seemed like the biggest difference from the user's view of the readLine routines (and I assumed the underlying implementation hadn't changed much).

If it is the slice, it gets a bit sticky:  Would I want to re-introduce the old interface even though it's clunky compared to the current, and slicing is a good way to express this?  Not really...  Is there more we could do to reduce the cost of slicing or peephole-optimize cases like this?  Maybe, though the effort is obviously greater.

With respect to the code size change / two calls per loop, we had a discussion about that [here](https://github.com/chapel-lang/chapel/pull/19937#pullrequestreview-998951729) and it looks like we agreed that it could be written with only one call per loop, but then that didn't happen?  I'm not sure whether that means it fell by the wayside and was forgotten, or whether it meant we couldn't do it (easily?) for some reason.  I feel like I tried to rewrite one version to use one readLine() per loop at some point and failed, but don't remember the details anymore...

**Comment by bradcray** [comment]
For reference, here's where our knucleotide results stood the last time I ran plots (and I don't think the site has been updated to include any new entries since then):

![image](https://user-images.githubusercontent.com/7536222/194427984-a8d885ed-43bb-49da-9c64-cb3853ca571b.png)

**Comment by mppf** [comment]
> If it is the slice, it gets a bit sticky: Would I want to re-introduce the old interface even though it's clunky compared to the current, and slicing is a good way to express this? Not really... Is there more we could do to reduce the cost of slicing or peephole-optimize cases like this? Maybe, though the effort is obviously greater.

I am wondering if the compound operators idea (https://github.com/chapel-lang/chapel/issues/17999#issuecomment-976579379) could cover this. Would it be possible for the IO module to define `proc readLine(_array.this(indices))` or something like that, that represents the combination of readLine + slicing, for optimization purposes?
